Python/Product/Analysis/LanguageServer/EditorFiles.cs

```diff
@@ -0,0 +1,237 @@
+                lock (_pendingChanges) {
+                    var notExpired = _pendingChanges
+                        .Where(p => p.textDocument.version.GetValueOrDefault() >= toVersion)
+                        .OrderBy(p => p.textDocument.version.GetValueOrDefault())
+                        .ToArray();
+
+                    _pendingChanges.Clear();
+                    if (notExpired.Any()) {
+                        next = notExpired.First();
+                        _pendingChanges.AddRange(notExpired.Skip(1));
+                    }
+                }
+                if (next.HasValue) {
+                    DidChangeTextDocument(next.Value, null);
+                }
+            } finally {
+                if (enqueueAction != null) {
+                    _server.TraceMessage($"Applied changes to {uri}");
+                    enqueueAction(doc);
+                }
+            }
+        }
+
+        public void UpdateParseDiagnostics(VersionCookie vc, Uri documentUri) {
+            List<PublishDiagnosticsEventArgs> diags = null;
+
+            lock (_lock) {
+                var last = _parseBufferDiagnostics;
+
+                foreach (var kv in vc.GetAllParts(documentUri)) {
+                    var part = _server.ProjectFiles.GetPart(kv.Key);
+                    if (!last.TryGetValue(part, out var lastVersion) || lastVersion.Version < kv.Value.Version || _ignoreDiagnosticsVersion) {
+                        last[part] = kv.Value;
+                            diags = diags ?? new List<PublishDiagnosticsEventArgs>();
+                            diags.Add(new PublishDiagnosticsEventArgs {
+                                uri = kv.Key,
+                                diagnostics = kv.Value.Diagnostics,
+                                _version = kv.Value.Version
+                            });
+                    }
+                }
+                _ignoreDiagnosticsVersion = false;
+                _lastReportedParseDiagnostics = diags ?? _lastReportedParseDiagnostics;
+
+                if (diags != null) {
+                    PublishDiagnostics(diags);
+                }
+            }
+        }
+
+        public void UpdateAnalysisDiagnostics(IProjectEntry projectEntry, int version) {
+            lock (_lock) {
+                var diags = _server.Analyzer.GetDiagnostics(projectEntry);
+                var settings = _server.Settings;
+
+                for (var i = 0; i < diags.Count; i++) {
+                    diags[i].severity = settings.analysis.GetEffectiveSeverity(diags[i].code, diags[i].severity);
+                }
+
+                var severity = settings.analysis.GetEffectiveSeverity(ErrorMessages.UnresolvedImportCode, DiagnosticSeverity.Warning);
+                var pythonProjectEntry = projectEntry as IPythonProjectEntry;
+                var parse = pythonProjectEntry?.GetCurrentParse();
+
+                if (parse != null && severity != DiagnosticSeverity.Unspecified) {
```

Comment:
Can you add a `TODO` comment to move this scan into the normal analysis process?

Reply:
Sure